### PR TITLE
Make sure Metadata is useful

### LINKED ISSUE #91: Handle "Metadata"
On the S3 HeadObjectOutput there is a parameter "Metadata" that is never set. 

I think, that all parameters on the object is the metadata..

### LINKED ISSUE #92: HeadObjectOutput::LastModified should be a \DateTimeImmutable
Like everything else marked as a "timestamp"

### PR DESCRIPTION
**Pull request by Nyholm**
This will fix #91 and fix #92 

~~It is basically an array of all properties. Have I understood this correctly?~~

### CODE REVIEW

**Comment by Nyholm** [comment]
> It is basically an array of all properties. Have I understood this correctly?

No. Metadata is extra properties you can set on an S3 object. They must begin with `x-amz-meta-`. 
https://docs.aws.amazon.com/AmazonS3/latest/user-guide/add-object-metadata.html

File: src/CodeGenerator/Generator/ApiGenerator.php [hunk 1]
```diff
@@ -719,15 +728,35 @@ private function resultClassPopulateResult(string $operationName, string $classN
...
             if (($member['location'] ?? null) !== 'header') {
                 $nonHeaders[$name] = $member;
 
                 continue;
             }
+            $headers[$name] = $member;
+
             $locationName = $member['locationName'] ?? $name;
-            $body .= "\$this->$name = \$headers['{$locationName}'];\n";
+            $body .= "\$this->$name = \$headers['{$locationName}'] ?? null;\n";
```
Comment (Nyholm) [inline src/CodeGenerator/Generator/ApiGenerator.php]: Header may not be defined, better use `??`

File: src/S3/Result/GetObjectOutput.php [hunk 2]
```diff
@@ -395,35 +395,37 @@ protected function populateResult(ResponseInterface $response, ?HttpClientInterf
...
+        $this->StorageClass = $headers['x-amz-storage-class'] ?? null;
+        $this->RequestCharged = $headers['x-amz-request-charged'] ?? null;
+        $this->ReplicationStatus = $headers['x-amz-replication-status'] ?? null;
+        $this->PartsCount = $headers['x-amz-mp-parts-count'] ?? null;
+        $this->TagCount = $headers['x-amz-tagging-count'] ?? null;
+        $this->ObjectLockMode = $headers['x-amz-object-lock-mode'] ?? null;
+        $this->ObjectLockRetainUntilDate = $headers['x-amz-object-lock-retain-until-date'] ?? null;
+        $this->ObjectLockLegalHoldStatus = $headers['x-amz-object-lock-legal-hold'] ?? null;
+        $this->Metadata = [
+        ];
```
Comment (jderusse) [inline src/S3/Result/GetObjectOutput.php]: is it expected to get an empty array?

Comment (Nyholm) [reply src/S3/Result/GetObjectOutput.php]: No. I've updated the PR.

**Comment by Nyholm** [comment]
Finally done. 

- All headers are lower key and multi dimension array. 
- I use `filter_var` on header result and body result
- I use `?? null` on headers to make sure they are null if not defined. 
- Support for S3 metadata. 

This will also fix #92

File: src/CodeGenerator/Generator/ApiGenerator.php [hunk 3]
```diff
@@ -667,6 +667,14 @@ private function resultClassAddProperties(string $baseNamespace, bool $root, ?ar
             if ('structure' === $memberShape['type']) {
                 $this->doGenerateResultClass($baseNamespace, $parameterType);
                 $parameterType = $baseNamespace . '\\' . $this->safeClassName($parameterType);
+            } elseif ('map' === $memberShape['type']) {
```
Comment (jderusse) [inline src/CodeGenerator/Generator/ApiGenerator.php]: The ApiGenerator contains several occurence of `if type == 'structure` or `if type == 'list'`. Shouldn't we consider `map` also?

Comment (Nyholm) [reply src/CodeGenerator/Generator/ApiGenerator.php]: We maybe should. I write an issue.

Comment (Nyholm) [reply src/CodeGenerator/Generator/ApiGenerator.php]: See #100

**Comment by Nyholm** [comment]
Thank you

**Pull request merged**